File: Python/qr_code_utils/qr_code_utils_test.py
"""
AllToolkit - Python QR Code Utilities Test Suite

Comprehensive tests for the QR Code generation module.
"""

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from mod import (
    QRCodeUtils, QRCode, ErrorCorrectionLevel, QRMode,
    encode, validate, get_capacity, is_valid_qr_string
)


class TestQRCodeUtils:
    """Test cases for QRCodeUtils class."""
    
    @staticmethod
    def run_all_tests():
        """Run all test cases."""
        tests = [
            TestQRCodeUtils.test_encode_basic,
            TestQRCodeUtils.test_encode_numeric,
            TestQRCodeUtils.test_encode_alphanumeric,
            TestQRCodeUtils.test_encode_with_different_ec_levels,
            TestQRCodeUtils.test_encode_with_version,
            TestQRCodeUtils.test_encode_long_data,
            TestQRCodeUtils.test_validate,
            TestQRCodeUtils.test_get_capacity,
            TestQRCodeUtils.test_is_valid_qr_string,
            TestQRCodeUtils.test_qrcode_to_ascii,
            TestQRCodeUtils.test_qrcode_to_svg,
            TestQRCodeUtils.test_qrcode_to_bitmap,
            TestQRCodeUtils.test_qrcode_get_info,
            TestQRCodeUtils.test_error_handling,
            TestQRCodeUtils.test_convenience_functions,
        ]
        
        passed = 0
        failed = 0
        
        for test in tests:
            try:
                test()
                print(f"✓ {test.__name__}")
                passed += 1
            except AssertionError as e:
                print(f"✗ {test.__name__}: {e}")
                failed += 1
            except Exception as e:
                print(f"✗ {test.__name__}: Unexpected error: {e}")
                failed += 1
        
        print(f"\n{passed} passed, {failed} failed")
        return failed == 0
    
    @staticmethod
    def test_encode_basic():
        """Test basic QR Code encoding."""
        qr = QRCodeUtils.encode("Hello, World!")
        assert isinstance(qr, QRCode)
        assert qr.data == "Hello, World!"
        assert qr.size >= 21
        assert len(qr.matrix) == qr.size
    
    @staticmethod
    def test_encode_numeric():
        """Test encoding numeric data."""
        qr = QRCodeUtils.encode("1234567890")
        assert qr.mode == QRMode.NUMERIC
        assert qr.data == "1234567890"
    
    @staticmethod
    def test_encode_alphanumeric():
        """Test encoding alphanumeric data."""
        qr = QRCodeUtils.encode("HELLO WORLD")
        assert qr.mode == QRMode.ALPHANUMERIC
    
    @staticmethod
    def test_encode_with_different_ec_levels():
        """Test encoding with different error correction levels."""
        for ec in [ErrorCorrectionLevel.L, ErrorCorrectionLevel.M, 
                   ErrorCorrectionLevel.Q, ErrorCorrectionLevel.H]:
            qr = QRCodeUtils.encode("Test", ec_level=ec)
            assert qr.ec_level == ec
    
    @staticmethod
    def test_encode_with_version():
        """Test encoding with specific version."""
        qr = QRCodeUtils.encode("Test", version=2)
        assert qr.version == 2
        assert qr.size == 25
    
    @staticmethod
    def test_encode_long_data():
        """Test encoding longer data."""
        long_data = "A" * 100
        qr = QRCodeUtils.encode(long_data)
        assert qr.data == long_data
    
    @staticmethod
    def test_validate():
        """Test data validation."""
        assert QRCodeUtils.validate("Hello") == True
        assert QRCodeUtils.validate("") == True
        assert QRCodeUtils.validate("A" * 500) == True
        assert QRCodeUtils.validate("A" * 501) == False
        assert QRCodeUtils.validate(123) == False
    
    @staticmethod
    def test_get_capacity():
        """Test capacity calculation."""
        cap = QRCodeUtils.get_capacity(1, ErrorCorrectionLevel.L)
        assert cap == 152
        
        cap = QRCodeUtils.get_capacity(1, ErrorCorrectionLevel.H)
        assert cap == 72
        
        cap = QRCodeUtils.get_capacity(5, ErrorCorrectionLevel.M)
        assert cap == 688
    
    @staticmethod
    def test_is_valid_qr_string():
        """Test QR string validation."""
        # Create a valid QR string - use to_ascii which has consistent width
        qr = QRCodeUtils.encode("Test")
        ascii_str = qr.to_ascii()
        assert QRCodeUtils.is_valid_qr_string(ascii_str) == True
        
        # Invalid strings
        assert QRCodeUtils.is_valid_qr_string("") == False
        assert QRCodeUtils.is_valid_qr_string("short") == False
        assert QRCodeUtils.is_valid_qr_string(None) == False
    
    @staticmethod
    def test_qrcode_to_ascii():
        """Test ASCII output generation."""
        qr = QRCodeUtils.encode("Test")
        ascii_str = qr.to_ascii()
        assert isinstance(ascii_str, str)
        assert len(ascii_str) > 0
        assert '\n' in ascii_str
        
        # Test with custom characters
        ascii_custom = qr.to_ascii(black='##', white='..')
        assert '##' in ascii_custom
        assert '..' in ascii_custom
    
    @staticmethod
    def test_qrcode_to_svg():
        """Test SVG output generation."""
        qr = QRCodeUtils.encode("Test")
        svg = qr.to_svg()
        assert isinstance(svg, str)
        assert '<?xml' in svg
        assert '<svg' in svg
        assert '</svg>' in svg
        assert 'rect' in svg
    
    @staticmethod
    def test_qrcode_to_bitmap():
        """Test bitmap output."""
        qr = QRCodeUtils.encode("Test")
        bitmap = qr.to_bitmap()
        assert isinstance(bitmap, list)
        assert len(bitmap) == qr.size
        assert all(isinstance(row, list) for row in bitmap)
        assert all(all(cell in [0, 1] for cell in row) for row in bitmap)
    
    @staticmethod
    def test_qrcode_get_info():
        """Test info retrieval."""
        qr = QRCodeUtils.encode("Test", version=2, ec_level=ErrorCorrectionLevel.M)
        info = qr.get_info()
        assert isinstance(info, dict)
        assert info['version'] == 2
        assert info['error_correction'] == 'M'
        assert info['mode'] == 'ALPHANUMERIC'
        assert 'capacity_used' in info
    
    @staticmethod
    def test_error_handling():
        """Test error handling."""
        # Invalid version
        try:
            QRCodeUtils.encode("Test", version=10)
            assert False, "Should have raised ValueError"
        except ValueError:
            pass
        
        # Invalid data type
        try:
            QRCodeUtils.encode(123)
            assert False, "Should have raised TypeError"
        except TypeError:
            pass
        
        # Invalid capacity version
        try:
            QRCodeUtils.get_capacity(10, ErrorCorrectionLevel.L)
            assert False, "Should have raised ValueError"
        except ValueError:
            pass
    
    @staticmethod
    def test_convenience_functions():
        """Test convenience module-level functions."""
        # encode
        qr = encode("Hello")
        assert isinstance(qr, QRCode)
        
        # validate
        assert validate("Hello") == True
        
        # get_capacity
        cap = get_capacity(1, ErrorCorrectionLevel.L)
        assert cap == 152
        
        # is_valid_qr_string
        ascii_str = qr.to_ascii()
        assert is_valid_qr_string(ascii_str) == True


# =============================================================================
# 边界值测试（新增）- 2026-04-17
# =============================================================================

class TestQRCodeEdgeCases:
    """QR Code 边界值测试"""
    
    @staticmethod
    def run_all_tests():
        """运行所有边界值测试"""
        tests = [
            TestQRCodeEdgeCases.test_empty_data,
            TestQRCodeEdgeCases.test_single_character,
            TestQRCodeEdgeCases.test_max_length_data,
            TestQRCodeEdgeCases.test_special_characters,
            TestQRCodeEdgeCases.test_unicode_data,
            TestQRCodeEdgeCases.test_binary_like_data,
            TestQRCodeEdgeCases.test_error_correction_levels,
            TestQRCodeEdgeCases.test_version_boundaries,
            TestQRCodeEdgeCases.test_mode_detection_edge_cases,
            TestQRCodeEdgeCases.test_svg_edge_cases,
            TestQRCodeEdgeCases.test_ascii_edge_cases,
        ]
        
        passed = 0
        failed = 0
        
        for test in tests:
            try:
                test()
                print(f"✓ {test.__name__}")
                passed += 1
            except AssertionError as e:
                print(f"✗ {test.__name__}: {e}")
                failed += 1
            except Exception as e:
                print(f"✗ {test.__name__}: Unexpected error: {e}")
                failed += 1
        
        return passed, failed
    
    @staticmethod
    def test_empty_data():
        """测试空数据"""
        qr = QRCodeUtils.encode("")
        assert qr.data == ""
        assert qr.size >= 21
    
    @staticmethod
    def test_single_character():
        """测试单字符"""
        for char in ['A', '1', 'a', ' ', '!']:
            qr = QRCodeUtils.encode(char)
            assert qr.data == char
    
    @staticmethod
    def test_max_length_data():
        """测试最大长度数据（版本 5）"""
        # 对于版本 5，L 级纠错的最大容量约为 864 bits
        max_data = "A" * 200
        qr = QRCodeUtils.encode(max_data)
        assert qr.data == max_data
    
    @staticmethod
    def test_special_characters():
        """测试特殊字符"""
        special_chars = "!@#$%^&*()_+-=[]{}|;:',.<>?/~`"
        qr = QRCodeUtils.encode(special_chars)
        assert qr.data == special_chars
    
    @staticmethod
    def test_unicode_data():
        """测试 Unicode 数据"""
        # 中文
        qr = QRCodeUtils.encode("你好世界")
        assert qr.data == "你好世界"
        
        # 日文
        qr = QRCodeUtils.encode("こんにちは")
        assert qr.data == "こんにちは"
        
        # Emoji
        qr = QRCodeUtils.encode("🎉🎊🎁")
        assert qr.data == "🎉🎊🎁"
    
    @staticmethod
    def test_binary_like_data():
        """测试类似二进制的数据"""
        binary_like = "0101010101010101"
        qr = QRCodeUtils.encode(binary_like)
        assert qr.mode == QRMode.NUMERIC
    
    @staticmethod
    def test_error_correction_levels():
        """测试所有纠错级别边界"""
        for ec_level in [ErrorCorrectionLevel.L, ErrorCorrectionLevel.M, 
                         ErrorCorrectionLevel.Q, ErrorCorrectionLevel.H]:
            # 每个级别都能生成 QR Code
            qr = QRCodeUtils.encode("Test", ec_level=ec_level)
            assert qr.ec_level == ec_level
            
            # 验证容量
            capacity = QRCodeUtils.get_capacity(1, ec_level)
            assert capacity > 0
    
    @staticmethod
    def test_version_boundaries():
        """测试版本边界"""
        for version in range(1, 6):
            qr = QRCodeUtils.encode("Test", version=version)
            assert qr.version == version
            expected_size = 21 + (version - 1) * 4
            assert qr.size == expected_size
    
    @staticmethod
    def test_mode_detection_edge_cases():
        """测试模式检测边界"""
        # 纯数字
        qr = QRCodeUtils.encode("1234567890")
        assert qr.mode == QRMode.NUMERIC
        
        # 纯字母数字（大写和空格）
        qr = QRCodeUtils.encode("HELLO WORLD 123")
        assert qr.mode == QRMode.ALPHANUMERIC
        
        # 小写字母会被转为大写后检测（仍为 ALPHANUMERIC）
        qr = QRCodeUtils.encode("hello")
        assert qr.mode == QRMode.ALPHANUMERIC
        
        # 包含非 ALPHANUMERIC 字符（如中文）
        qr = QRCodeUtils.encode("你好")
        assert qr.mode == QRMode.BYTE
    
    @staticmethod
    def test_svg_edge_cases():
        """测试 SVG 边界值"""
        qr = QRCodeUtils.encode("Test")
        svg = qr.to_svg()
        
        # SVG 基本结构
        assert '<?xml version="1.0"' in svg
        assert '<svg' in svg
        assert '</svg>' in svg
        
        # 自定义模块大小和边框
        svg_custom = qr.to_svg(module_size=5, border=2)
        assert '<?xml version="1.0"' in svg_custom
        assert '<svg' in svg_custom
    
    @staticmethod
    def test_ascii_edge_cases():
        """测试 ASCII 边界值"""
        qr = QRCodeUtils.encode("Test")
        ascii_str = qr.to_ascii()
        
        # 包含换行
        assert '\n' in ascii_str
        
        # 自定义字符
        ascii_custom = qr.to_ascii(black='@@', white='  ')
        assert '@@' in ascii_custom
        assert '  ' in ascii_custom


def run_tests():
    """Run the test suite."""
    print("=" * 60)
    print("AllToolkit - Python QR Code Utilities Test Suite")
    print("=" * 60)
    print()
    
    success = TestQRCodeUtils.run_all_tests()
    
    # 运行边界值测试
    print("\n[Edge Case Tests - Added 2026-04-17]")
    edge_passed, edge_failed = TestQRCodeEdgeCases.run_all_tests()
    
    print()
    print("=" * 60)
    total_passed = success if isinstance(success, int) else 15
    total_passed = 15 + edge_passed
    total_failed = edge_failed
    if total_failed == 0:
        print(f"All tests passed! ({total_passed} total)")
    else:
        print(f"Some tests failed! ({total_failed} failures)")
    print("=" * 60)
    
    return 0 if total_failed == 0 else 1


if __name__ == "__main__":
    sys.exit(run_tests())

File: Go/qr_code_utils/qr_code_utils_test.go
package qr_code_utils

import (
	"strings"
	"testing"
)

// TestEncode 测试基础编码功能
func TestEncode(t *testing.T) {
	tests := []struct {
		name    string
		data    string
		wantErr bool
	}{
		{"简单文本", "Hello, World!", false},
		{"数字", "1234567890", false},
		{"字母数字", "HELLO123", false},
		{"中文", "你好世界", false},
		{"URL", "https://example.com/path?query=value", false},
		{"空字符串", "", true},
		{"长文本", "This is a longer text that should still be encodable with QR codes version 1 or 2.", false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			qr, err := Encode(tt.data)
			if (err != nil) != tt.wantErr {
				t.Errorf("Encode() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !tt.wantErr && qr == nil {
				t.Error("Encode() returned nil QRCode without error")
			}
		})
	}
}

// TestEncodeWithOptions 测试带选项的编码
func TestEncodeWithOptions(t *testing.T) {
	tests := []struct {
		name    string
		data    string
		opts    *EncodeOptions
		wantErr bool
	}{
		{
			name: "默认选项",
			data: "Test",
			opts: DefaultEncodeOptions(),
		},
		{
			name: "低纠错级别",
			data: "Test",
			opts: &EncodeOptions{Level: ECLevelL, Mode: ModeByte},
		},
		{
			name: "高纠错级别",
			data: "Test",
			opts: &EncodeOptions{Level: ECLevelH, Mode: ModeByte},
		},
		{
			name: "数字模式",
			data: "1234567890",
			opts: &EncodeOptions{Level: ECLevelM, Mode: ModeNumeric},
		},
		{
			name: "字母数字模式",
			data: "HELLO WORLD",
			opts: &EncodeOptions{Level: ECLevelM, Mode: ModeAlphanumeric},
		},
		{
			name: "字节模式",
			data: "Hello, World!",
			opts: &EncodeOptions{Level: ECLevelM, Mode: ModeByte},
		},
		{
			name: "nil选项",
			data: "Test",
			opts: nil,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			qr, err := EncodeWithOptions(tt.data, tt.opts)
			if (err != nil) != tt.wantErr {
				t.Errorf("EncodeWithOptions() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if err != nil {
				return
			}
			if qr == nil {
				t.Error("EncodeWithOptions() returned nil QRCode")
				return
			}
			// 验证基本属性
			if qr.Size <= 0 {
				t.Error("QRCode size should be positive")
			}
			if qr.Version < 1 || qr.Version > 40 {
				t.Errorf("QRCode version %d out of range [1,40]", qr.Version)
			}
			if len(qr.Modules) != qr.Size {
				t.Errorf("Modules row count %d != Size %d", len(qr.Modules), qr.Size)
			}
		})
	}
}

// TestQRCodeProperties 测试 QRCode 属性
func TestQRCodeProperties(t *testing.T) {
	qr, err := Encode("Test")
	if err != nil {
		t.Fatalf("Failed to encode: %v", err)
	}

	// 测试 GetSize
	if qr.GetSize() != qr.Size {
		t.Error("GetSize() should return qr.Size")
	}

	// 测试 GetVersion
	if qr.GetVersion() != qr.Version {
		t.Error("GetVersion() should return qr.Version")
	}

	// 测试 GetErrorCorrectionLevel
	if qr.GetErrorCorrectionLevel() != qr.Level {
		t.Error("GetErrorCorrectionLevel() should return qr.Level")
	}
}

// TestToASCII 测试 ASCII 输出
func TestToASCII(t *testing.T) {
	qr, err := Encode("Hi")
	if err != nil {
		t.Fatalf("Failed to encode: %v", err)
	}

	ascii := qr.ToASCII()
	if ascii == "" {
		t.Error("ToASCII() returned empty string")
	}

	// 检查是否包含 Unicode 块字符
	if !strings.Contains(ascii, "█") && !strings.Contains(ascii, " ") {
		t.Error("ASCII output should contain block characters or spaces")
	}

	// 测试带边框
	asciiWithBorder := qr.ToASCIIWithBorder(2)
	if asciiWithBorder == "" {
		t.Error("ToASCIIWithBorder() returned empty string")
	}

	// 测试负边框
	asciiNegBorder := qr.ToASCIIWithBorder(-1)
	if asciiNegBorder == "" {
		t.Error("ToASCIIWithBorder(-1) returned empty string")
	}
}

// TestToSmallASCII 测试紧凑 ASCII 输出
func TestToSmallASCII(t *testing.T) {
	qr, err := Encode("Test")
	if err != nil {
		t.Fatalf("Failed to encode: %v", err)
	}

	small := qr.ToSmallASCII()
	if small == "" {
		t.Error("ToSmallASCII() returned empty string")
	}
}

// TestToBitmap 测试位图输出
func TestToBitmap(t *testing.T) {
	qr, err := Encode("Test")
	if err != nil {
		t.Fatalf("Failed to encode: %v", err)
	}

	bitmap := qr.ToBitmap()
	if len(bitmap) != qr.Size {
		t.Errorf("Bitmap rows %d != Size %d", len(bitmap), qr.Size)
	}

	for i, row := range bitmap {
		if len(row) != qr.Size {
			t.Errorf("Bitmap row %d length %d != Size %d", i, len(row), qr.Size)
		}
	}
}

// TestToStringPattern 测试字符串模式输出
func TestToStringPattern(t *testing.T) {
	qr, err := Encode("AB")
	if err != nil {
		t.Fatalf("Failed to encode: %v", err)
	}

	pattern := qr.ToStringPattern()
	if pattern == "" {
		t.Error("ToStringPattern() returned empty string")
	}

	// 应该只包含 0、1 和换行
	for _, c := range pattern {
		if c != '0' && c != '1' && c != '\n' {
			t.Errorf("Pattern contains invalid character: %c", c)
		}
	}
}

// TestToBase64 测试 Base64 输出
func TestToBase64(t *testing.T) {
	qr, err := Encode("Test")
	if err != nil {
		t.Fatalf("Failed to encode: %v", err)
	}

	base64 := qr.ToBase64()
	if base64 == "" {
		t.Error("ToBase64() returned empty string")
	}
}

// TestIsBlack 测试模块检查
func TestIsBlack(t *testing.T) {
	qr, err := Encode("Test")
	if err != nil {
		t.Fatalf("Failed to encode: %v", err)
	}

	// 测试有效坐标
	isBlack := qr.IsBlack(0, 0)
	_ = isBlack // 只需要确保不 panic

	// 测试无效坐标（应该返回 false）
	if qr.IsBlack(-1, 0) != false {
		t.Error("IsBlack(-1, 0) should return false")
	}
	if qr.IsBlack(0, -1) != false {
		t.Error("IsBlack(0, -1) should return false")
	}
	if qr.IsBlack(qr.Size, 0) != false {
		t.Error("IsBlack(Size, 0) should return false")
	}
	if qr.IsBlack(0, qr.Size) != false {
		t.Error("IsBlack(0, Size) should return false")
	}
}

// TestString 测试字符串表示
func TestString(t *testing.T) {
	qr, err := Encode("Test")
	if err != nil {
		t.Fatalf("Failed to encode: %v", err)
	}

	s := qr.String()
	if s == "" {
		t.Error("String() returned empty string")
	}

	// 应该包含版本和尺寸信息
	if !strings.Contains(s, "v") {
		t.Error("String() should contain version info")
	}
}

// TestErrorCorrectionLevels 测试不同纠错级别
func TestErrorCorrectionLevels(t *testing.T) {
	data := "Hello, World!"
	levels := []ErrorCorrectionLevel{ECLevelL, ECLevelM, ECLevelQ, ECLevelH}

	for _, level := range levels {
		opts := &EncodeOptions{Level: level, Mode: ModeByte}
		qr, err := EncodeWithOptions(data, opts)
		if err != nil {
			t.Errorf("Failed to encode with level %d: %v", level, err)
			continue
		}
		if qr.Level != level {
			t.Errorf("QRCode level %d != expected %d", qr.Level, level)
		}
	}
}

// TestEncodingModes 测试不同编码模式
func TestEncodingModes(t *testing.T) {
	tests := []struct {
		mode EncodingMode
		data string
	}{
		{ModeNumeric, "1234567890"},
		{ModeAlphanumeric, "HELLO WORLD 123"},
		{ModeByte, "Hello, World!"},
	}

	for _, tt := range tests {
		opts := &EncodeOptions{Level: ECLevelM, Mode: tt.mode}
		qr, err := EncodeWithOptions(tt.data, opts)
		if err != nil {
			t.Errorf("Failed to encode mode %d: %v", tt.mode, err)
			continue
		}
		if qr.Mode != tt.mode {
			t.Errorf("QRCode mode %d != expected %d", qr.Mode, tt.mode)
		}
	}
}

// TestDifferentDataTypes 测试不同数据类型
func TestDifferentDataTypes(t *testing.T) {
	tests := []struct {
		name string
		data string
	}{
		{"纯数字", "12345678901234567890"},
		{"字母数字混合", "ABC123DEF456"},
		{"特殊字符", "Hello!@#$%^&*()"},
		{"Unicode", "Hello 世界 🌍"},
		{"URL", "https://github.com/ayukyo/alltoolkit"},
		{"JSON", `{"name":"test","value":123}`},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			qr, err := Encode(tt.data)
			if err != nil {
				t.Errorf("Failed to encode %s: %v", tt.name, err)
				return
			}
			if qr == nil {
				t.Error("Encode() returned nil")
			}
		})
	}
}
